feat: refactoring of text message renderer, fix interactive elements focus order, make chat user avatar keyboard accessible

arjita-mitra opened the pull request:
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - refactor text message renderer and fix focus order of all interactive elements

- The **PR Description**
  - refactor text message renderer
  -  fix focus order of all interactive elements
  - handle media elements focus
  - make avatar keyboard accessible
----

# What's new in this PR?

### Issues

  - refactor text message rendere
  -  fix focus order of all interactive elements
  - handle media elements focus
  - make avatar keyboard accessible

### Testing

#### How to Test

- play screen reader and use tab/ arrow keys

----
##### References
1. https://wearezeta.atlassian.net/browse/ACC-268

arjita-mitra committed:
fix: code review fix and chat area focus improvements(ACCC-268)

handle all interactive elements focus with a function
hanlde media elements seperately as they have a loading state
make Avatar keyboard accessible
handle reply messsage focus

arjita-mitra committed:
fix: replace clickhandler with my new function(ACC-268)

arjita-mitra committed:
refactor: create a message event function(ACC-268)

atomrc commented on src/script/components/Avatar.tsx:
```diff
@@ -144,7 +154,7 @@ const Avatar: FC<AvatarProps> = ({
       noBadge={noBadge}
       noFilter={noFilter}
-      onClick={clickHandler}
+      handleAvatarInteraction={handleAvatarInteraction}
```
```suggestion
      onAvatarInteraction={handleAvatarInteraction}
```

The component should not use imperative style for callbacks. It doesn't really know what those callbacks actually do, so we could use the `on` convention (`onClick`, `onKeyboard`, `onAvatarInteraction` ...)

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx:
```diff
@@ -67,97 +67,48 @@ export const TextMessageRenderer: FC<TextMessageRendererProps> = ({
+    const markdownLinkTargets = containerRef.current && [...containerRef.current.querySelectorAll('[data-md-link]')];
     const hasMentions = asset && asset.mentions().length;
     const msgMention = hasMentions
       ? containerRef.current && [...containerRef.current.querySelectorAll('.message-mention')]
```
```suggestion
      ? [...containerRef.current.querySelectorAll('.message-mention')]
```

You don't need to check for containerRef here, you are already doing it earlier in the code :)

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx:
```diff
@@ -67,97 +67,48 @@ export const TextMessageRenderer: FC<TextMessageRendererProps> = ({
       return undefined;
     }
     const emailLinks = containerRef.current && [...containerRef.current.querySelectorAll('[data-email-link]')];
```
```suggestion
    const emailLinks =  [...containerRef.current.querySelectorAll('[data-email-link]')];
```

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx:
```diff
@@ -67,97 +67,48 @@ export const TextMessageRenderer: FC<TextMessageRendererProps> = ({
     const emailLinks = containerRef.current && [...containerRef.current.querySelectorAll('[data-email-link]')];
-    const linkTargets = containerRef.current && [...containerRef.current.querySelectorAll('a[data-md-link]')];
-    const msgLinkTargets = containerRef.current && [...containerRef.current.querySelectorAll('[data-uie-name]')];
+    const markdownLinkTargets = containerRef.current && [...containerRef.current.querySelectorAll('[data-md-link]')];
```
```suggestion
    const markdownLinkTargets = [...containerRef.current.querySelectorAll('[data-md-link]')];
```

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/index.tsx:
```diff
@@ -70,9 +71,11 @@ const ContentAsset = ({
-              msgClass={`text ${includesOnlyEmojis(asset.text) ? 'text-large' : ''} ${
-                status === StatusType.SENDING ? 'text-foreground' : ''
-              } ${isObfuscated ? 'ephemeral-message-obfuscated' : ''}`}
+              msgClass={cx('text', {
+                'text-foreground': status === StatusType.SENDING,
+                'text-large': includesOnlyEmojis(asset.text),
+                'ephemeral-message-obfuscated': isObfuscated,
+              })}
```
👍

arjita-mitra committed:
Update src/script/components/Avatar.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

arjita-mitra committed:
Update src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

arjita-mitra committed:
Update src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

arjita-mitra committed:
Update src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx

Co-authored-by: Thomas Belin <thomasbelin4@gmail.com>

atomrc commented on src/script/util/util.ts:
```diff
@@ -422,3 +422,35 @@ export const removeAnimationsClass = (element: HTMLElement | null) => {
+      element.setAttribute('tabindex', isCurrentConversationFocused ? '0' : '-1');
+    });
+  }
+};
```
This doesn't really belong in this file. 

maybe a new file in `/script/ui/focusUtils.ts` would be better. WDYT?

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx:
```diff
@@ -66,98 +66,49 @@ export const TextMessageRenderer: FC<TextMessageRendererProps> = ({
+    const handleMsgEvent = (event: Event) => {
       return event.type === 'keydown'
         ? handleKeyEvent(event as KeyboardEvent, 'mention')
         : onMessageClick(asset, event as MouseEvent, 'mention');
```
It seems you are always sending a `mention` click even when clicking on a link or an email. Am I wrong?

atomrc commented on src/script/components/MessagesList/Message/ContentMessage/asset/TextMessageRenderer.tsx:
```diff
@@ -66,98 +66,49 @@ export const TextMessageRenderer: FC<TextMessageRendererProps> = ({
+      removeEventListener(markdownLinkTargets, 'markdownLink');
+      removeEventListener(msgMention, 'mention');
     };
   }, [onMessageClick, asset, isCurrentConversationFocused, containerRef]);
```
```suggestion
  }, [onMessageClick, asset, isCurrentConversationFocused]);
```

Refs do not need to be added to react dependencies as they do not trigger re-renders

arjita-mitra committed:
fix: code review comments and refactoring

detect message type from current target as we can't pass element type to avoid creating a new function reference
handle Quote message focus state
change mention message element type from span to div with role button  as its an interactive element
lint error
test fix
